lib/ClangImporter/ImportDecl.cpp

```diff
@@ -8411,17 +8418,49 @@ ClangImporter::Implementation::importSwiftAttrAttributes(Decl *MappedDecl) {
         continue;
       }
 
-      // Dig out a source file we can use for parsing.
-      auto &sourceFile = getClangSwiftAttrSourceFile(
-          *MappedDecl->getDeclContext()->getParentModule(),
-          swiftAttr->getAttribute());
+      bool cached = true;
+      while (true) {
+        // Dig out a source file we can use for parsing.
+        auto &sourceFile = getClangSwiftAttrSourceFile(
+            *MappedDecl->getDeclContext()->getParentModule(),
+            swiftAttr->getAttribute(),
+            cached);
+
+        auto topLevelDecls = sourceFile.getTopLevelDecls();
+
+        // If we're using the cached version, check whether we can correctly
+        // clone the attribute.
+        if (cached) {
+          bool hasNonclonableAttribute = false;
+          for (auto decl : topLevelDecls) {
+            if (hasNonclonableAttribute)
+              break;
+
+            for (auto attr : decl->getAttrs()) {
+              if (auto customAttr = dyn_cast<CustomAttr>(attr)) {
```

Comment:
We probably want to also capture all subclasses with `UNIMPLEMENTED_CLONE` as well. Do we want the implementation of `CustomAttr::clone()` to assert if the argument list isn't empty?

Reply:
Yes, we can do that. Actually, I can avoid the clone here altogether with a different approach that's more similar to how member-attribute macros work.